// astroturf/traverse.js
"use strict";

exports.__esModule = true;
exports.default = traverse;
var _path = _interopRequireDefault(require("path"));
var _core = require("@babel/core");
var _plugin = _interopRequireDefault(require("./plugin"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function traverse(source, filename, opts) {
  const extname = _path.default.extname(filename);
  return (0, _core.transformSync)(source, {
    filename,
    babelrc: false,
    configFile: false,
    code: false,
    ast: false,
    plugins: [[_plugin.default, opts]],
    sourceType: 'unambiguous',
    parserOpts: {
      allowImportExportEverywhere: true,
      allowReturnOutsideFunction: true,
      allowSuperOutsideMethod: true,
      sourceType: 'unambiguous',
      sourceFilename: filename,
      plugins: ['jsx', extname === '.ts' || extname === '.tsx' ? 'typescript' : 'flow', 'doExpressions', 'objectRestSpread', 'decorators-legacy', 'classProperties', 'classPrivateProperties', 'classPrivateMethods', 'exportDefaultFrom', 'exportNamespaceFrom', 'asyncGenerators', 'functionBind', 'functionSent', 'dynamicImport', 'numericSeparator', 'optionalChaining', 'importMeta', 'bigInt', 'optionalCatchBinding', 'throwExpressions', ['pipelineOperator', {
        proposal: 'minimal'
      }], 'nullishCoalescingOperator']
    }
  });
}
